File: lib/core/BUTableCellView.m
//
//  BUTableCellView.m
//
//
//  Created by Neil Edwards on 12/08/2011.
//  Copyright 2011 Buffer. All rights reserved.
//

#import "BUTableCellView.h"

@implementation BUTableCellView



//
/***********************************************
 * @description			if this is called, the tableview will automatically create a new nib instance if required
 ***********************************************/
//
+(void)cacheCellForTableView:(UITableView*)tableView fromNib:(UINib *)nib{
	
	NSString *cellID = [self cellIdentifier];
	
	[tableView registerNib:nib forCellReuseIdentifier:cellID];
	
}


+ (id)cellForTableView:(UITableView *)tableView fromNib:(UINib *)nib {
    NSString *cellID = [self cellIdentifier];
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellID];
	
    if (cell == nil) {
        NSArray *nibObjects = [nib instantiateWithOwner:nil options:nil];
		
        NSAssert2(([nibObjects count] > 0) && 
                  [[nibObjects objectAtIndex:0] isKindOfClass:[self class]],
                  @"Nib '%@' does not appear to contain a valid %@", 
                  [self nibName], NSStringFromClass([self class]));
        
        cell = [nibObjects objectAtIndex:0];
    }
    
    return cell;    
}



- (void)awakeFromNib {
    
    [super awakeFromNib];
    [self initialise];
    
}


-(void)initialise{}

-(void)populate{}



-(IBAction)cellButtonWasSelected:(id)sender{}


-(void)sendCellButtonNotification:(NSDictionary*)dict{
	
	if(dict!=nil)
		[[NSNotificationCenter defaultCenter] postNotificationName:BUCELLNOTIFICATION object:nil userInfo:
		 [NSDictionary dictionaryWithObjectsAndKeys:[self className],@"type",dict,EVENTDICT,nil]];
		
}



+ (NSString *)cellIdentifier {
    return [NSString stringWithFormat:@"%@Identifier",[self nibName]];
}


#pragma row Heights
+(int)rowHeight{
	return STANDARDCELLHEIGHT;
}
+(NSNumber*)heightForCellWithDataProvider:(id)data{
    return [NSNumber numberWithInt:[self rowHeight]];
}


#pragma mark UINib support

+ (NSString *)className{
    return NSStringFromClass([self class]);
}
- (NSString *)className{
    return NSStringFromClass([self class]);
}

+ (UINib *)nib {
    NSBundle *classBundle = [NSBundle bundleForClass:[self class]];
    return [UINib nibWithNibName:[self nibName] bundle:classBundle];
}

+ (NSString *)nibName {
    return [[self className] stringByReplacingOccurrencesOfString:@"View" withString:@""];
}

@end

File: CycleStreets4.0/src/extensions/BUTableCellView.swift
//
//  BUTableCellView.swift
//
//  Created by Neil Edwards on 24/10/2015.
//

import Foundation
import UIKit


public typealias BUTableCellViewButtonBlock=(String,AnyObject)->()


protocol BUTableCellViewCore:class{

    func initialise()
    func populate()
    
    static func rowHeight()->Int

}

protocol BUTableCellViewButtonEvent:class{

    var cellButtonEventBlock:BUTableCellViewButtonBlock? { get set }

}


extension BUTableCellViewButtonEvent where Self:UITableViewCell{
    
    
    
    func cellButtonWasSelected(_ sender:AnyObject){}
    
    
    func sendCellButtonBlockWithEvent(_ event:String,dataProvider:AnyObject){
        
        if let _=cellButtonEventBlock{
            
            cellButtonEventBlock?(event,dataProvider)
            
        }
        
    }
    
}


extension BUTableCellViewCore where Self:UITableViewCell{
    
    
    func awakeFromNib(){
        self.awakeFromNib()
        self.initialise()
    }

    
    static func cellIdentifier()->String{
        return String(format: "%@Identifier", self.nibName())
    }
    
    static func rowHeight()->Int{
        return 44
    }
    
}



//class BUTableCellView:UITableViewCell,BUTableCellViewCore{
//    
//    
//    //MARK: - Instance
//    
//    override func awakeFromNib(){
//        
//        super.awakeFromNib()
//        self.initialise()
//        
//    }
//    
//    
//    
//}
